scalafix-reflect/src/main/scala-3/scalafix/internal/reflect/RuleCompiler.scala

```diff
@@ -0,0 +1,78 @@
+package scalafix.internal.reflect
+
+import dotty.tools.dotc.Compiler
+import dotty.tools.dotc.Run
+import dotty.tools.dotc.core.Contexts.FreshContext
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.reporting.StoreReporter
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.io.AbstractFile
+import dotty.tools.io.Directory
+import dotty.tools.io.PlainDirectory
+import dotty.tools.io.VirtualFile
+import dotty.tools.io.VirtualDirectory
+import dotty.tools.repl.AbstractFileClassLoader
+
+import metaconfig.Configured
+import metaconfig.Input
+import metaconfig.ConfError
+import metaconfig.Position
+
+import java.io.File
+class RuleCompiler(
+    classpath: String,
+    targetDirectory: Option[File] = None
+) {
+  private val output = targetDirectory match {
+    case Some(file) => new PlainDirectory(new Directory(file.toPath))
+    case None => new VirtualDirectory("(memory)")
+  }
+  private val settings =
+    "-unchecked" :: "-deprecation" :: "-classpath" :: classpath :: Nil
+  private val driver = new InteractiveDriver(settings)
+  private val reporter: StoreReporter = new StoreReporter()
+  private var ctx: FreshContext = driver.currentCtx.fresh
+  ctx = ctx
+    .setReporter(reporter)
+    .setSetting(ctx.settings.outputDir, output)
+    .setSetting(ctx.settings.classpath, classpath)
+
+  private val compiler: Compiler = new Compiler()
+
+  def compile(input: Input): Configured[ClassLoader] = {
+    reporter.removeBufferedMessages(using ctx)
+    val run: Run = compiler.newRun(using ctx)
+
+    val file: AbstractFile = input match {
+      case Input.File(path, _) => AbstractFile.getFile(input.path)
+      case Input.VirtualFile(path, _) =>
+        VirtualFile(input.path, input.text.getBytes())
+      case _ => throw RuntimeException("Invalid Input file")
+    }
+
+    run.compileSources(
+      List(new SourceFile(file, input.chars))
+    )
+
+    val errors = reporter.allErrors.map(error =>
+      ConfError
+        .message(error.getMessage)
+    )
+    if (!errors.isEmpty)
+      errors :+ ConfError.message(
+        "Error compiling rule(s) from source using Scala 3 compiler; " +
+          "to use the Scala 2.x compiler instead, use the corresponding " +
+          "scalafix-cli artifact or force scalafixScalaBinaryVersion " +
+          "to 2.x in your build tool"
```

Comment:
I don't manage to get that error when trying to compile source rules with quasiquotes with scalafix-cli_3

```
cs launch -r ivy2local ch.epfl.scala:scalafix-cli_3.1.3:0.10.4+22-b8cd292e+20221018-1241-SNAPSHOT -M scalafix.cli.Cli -- --rules=https://raw.githubusercontent.com/scalacenter/scalafix-named-literal-arguments/fb9f65ee9f09f167651b3d9bfe94a9be76818759/scalafix/rules/src/main/scala/fix/Namedliteralarguments_v1.scala
error: 2 errors
[E0]  Any is not a valid result type of an unapply method of an extractor.
[E1] Scala 2 macro cannot be used in Dotty. See https://dotty.epfl.ch/docs/reference/dropped-features/macros.html
To turn this error into a warning, pass -Xignore-scala2-macros to the compiler
```

Reply:
Added https://github.com/scalacenter/scalafix/pull/1650/commits/d15af9a0a1eb545819d0fb53ae3306b2e90313a4
```
cs launch -r ivy2local ch.epfl.scala:scalafix-cli_3.1.3:0.10.4+23-d15af9a0+20221018-1320-SNAPSHOT -M scalafix.cli.Cli -- --rules=https://raw.githubusercontent.com/scalacenter/scalafix-named-literal-arguments/fb9f65ee9f09f167651b3d9bfe94a9be76818759/scalafix/rules/src/main/scala/fix/Namedliteralarguments_v1.scala
error: 3 errors
[E0]  Any is not a valid result type of an unapply method of an extractor.
[E1] Scala 2 macro cannot be used in Dotty. See https://dotty.epfl.ch/docs/reference/dropped-features/macros.html
To turn this error into a warning, pass -Xignore-scala2-macros to the compiler
[E2] Error compiling rule(s) from source using Scala 3 compiler; to use the Scala 2.x compiler instead, use the corresponding scalafix-cli artifact or force scalafixScalaBinaryVersion to 2.x in your build tool
```